Custom shader for new backends

yshui opened the pull request:
Creating a new PR because I want to preserve @tryone144's original PR (#507) as a reference.

tryone144 committed:
win: keep track of window specific shader in `managed_win`

tryone144 committed:
backend: gl_common: add `time` uniform to (custom) window shader

yshui committed:
backend: add custom shader image property

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: gl: implement create_shader/destroy_shader

yshui committed:
config: add options `window-shader-fg` and `window-shader-fg-rules`

Used for setting custom window shaders and rules for choosing custom
window shaders.

Added a "c2_userdata_free" parameter to c2_list_free, so allocated userdata
stored in nodes can be freed.

Signed-off-by: Bernd Busse <bernd@busse-net.de>
Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: add shader related APIs

Current this API can only create shader from a single source file. This
could be changed in the future.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: gl: handle custom shader image property

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: gl: improve error reporting when creating shaders

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: init/deinit shaders when init/deinit-ing the backend

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
This should be ready for prime time. Only thing left is to document the shader interface.

Can you have a look, @tryone144 ?

yshui commented:
Ah, I forgot to set projection for custom shaders.

yshui commented:
Also, the time parameter is kinda useless unless we have an option to force picom to redraw every frame.

I propose adding a `force-redraw = bool;` option and a `force-redraw-rule = [];` option.

tryone144 commented:
> Can you have a look, @tryone144 ?

I will try to get around to this week.

> Also, the time parameter is kinda useless unless we have an option to force picom to redraw every frame.
> 
> I propose adding a `force-redraw = bool;` option and a `force-redraw-rule = [];` option.

I initially though about force-redrawing every window with a custom shader, but that is overkill performance wise. And I don't think autodetecting the use of `time` is reliable enough.
`force-redraw` and `force-redraw-rule` should be fine. Maybe even add a `force-blend-rule` option while we are at it?
These options however, leave the user responsible to ensure correct behaviour of their shaders (for animations and transparency).

yshui commented:
hmm, actually i think we can detect use of `time` automatically. `glGetUniformLocation` would return -1 if the variable isn't used.

tryone144 commented on src/picom.c:
```diff
@@ -2174,6 +2273,7 @@ static void session_destroy(session_t *ps) {
 	c2_list_free(&ps->o.paint_blacklist);
 	c2_list_free(&ps->o.unredir_if_possible_blacklist);
 	c2_list_free(&ps->o.rounded_corners_blacklist);
+	c2_list_free(&ps->o.window_shader_fg_rules);
```
We leak the filename allocated by `realpath()` in the `load_auxiliary_file*()` functions, that is stored as the userdata in the rule.

I suggest the following patch:
```diff
diff --git a/src/c2.c b/src/c2.c
index e557f29..437544f 100644
--- a/src/c2.c
+++ b/src/c2.c
@@ -1693,3 +1693,10 @@ bool c2_list_foreach(const c2_lptr_t *condlist, c2_list_foreach_cb_t cb, void *d
 void *c2_list_get_data(const c2_lptr_t *condlist) {
 	return condlist->data;
 }
+
+/// Free user data stored in a condition.
+void c2_list_free_data(const c2_lptr_t *condlist) {
+	if (condlist->data) {
+		free(condlist->data);
+	}
+}
diff --git a/src/c2.h b/src/c2.h
index 1f3390b..1c0a273 100644
--- a/src/c2.h
+++ b/src/c2.h
@@ -30,6 +30,8 @@ typedef bool (*c2_list_foreach_cb_t)(const c2_lptr_t *cond, void *data);
 bool c2_list_foreach(const c2_lptr_t *list, c2_list_foreach_cb_t cb, void *data);
 /// Return user data stored in a condition.
 void *c2_list_get_data(const c2_lptr_t *condlist);
+/// Free user data stored in a condition.
+void c2_list_free_data(const c2_lptr_t *condlist);
 
 /**
  * Destroy a condition list.
diff --git a/src/picom.c b/src/picom.c
index 047f5e7..7bed6b1 100644
--- a/src/picom.c
+++ b/src/picom.c
@@ -1604,6 +1604,11 @@ static bool load_shader_source_for_condition(const c2_lptr_t *cond, void *data)
 	return load_shader_source(data, c2_list_get_data(cond));
 }
 
+static bool free_shader_source_for_condition(const c2_lptr_t *cond, void *data attr_unused) {
+	c2_list_free_data(cond);
+	return false;
+}
+
 /**
  * Initialize a session.
  *
@@ -2273,6 +2278,9 @@ static void session_destroy(session_t *ps) {
 	c2_list_free(&ps->o.paint_blacklist);
 	c2_list_free(&ps->o.unredir_if_possible_blacklist);
 	c2_list_free(&ps->o.rounded_corners_blacklist);
+
+	// Free window shader rules
+	c2_list_foreach(ps->o.window_shader_fg_rules, free_shader_source_for_condition, NULL);
 	c2_list_free(&ps->o.window_shader_fg_rules);
 
 	// Free tracked atom list
 ```
Alternatively, we can change the opacity-rule to use an allocated `long` for its parameter and always free the userdata in `c2_list_free()`. We currently cannot blindly `free()` the userdata because the opacity-rules use the `void *` as storage for a `long`.

yshui replied:
I thought `c2_list_free` would free the user data :(. Turns out the user data pointer never stored allocated memory until now.

tryone144 reviewed (changes requested):
This refactor looks good. I just found some left over definitions from my code that aren't needed anymore and two memory leaks across backend-/session-destroy.

---
Regarding `time` and force-redraw:
The `time` uniform is part of `win_shader_t` i.e. the internal data of the backend. This is hidden behind the `backend_shader` pointer in `struct shader_info`.
The `create_shader()` backend operation could return a bool to indicate animation and we save that in the `shader_info` so we can check the window's attached shader whether to force-redraw.
If you can think of a more clean solution with less binding into the backend, that would be great.

Additionally, we could blend windows that have a non-default shader set, instead of relying on the user to set the `force-win-blend` option. Are there use-cases for this option apart from custom shaders?

yshui commented:
@tryone144 Thanks for the review! I will try to get to it ASAP

***
> The create_shader() backend operation could return a bool

Yes. I am also thinking maybe add a `get_shader_property` function with the only property being `ANIMATED` for now. I think this API is conceptually nicer, but I don't know if it's overkill. What other properties could a shader have?

yshui commented on src/backend/backend.c:
```diff
@@ -383,6 +383,12 @@ void paint_all_new(session_t *ps, struct managed_win *t, bool ignore_damage) {
 				    w->win_image, &border_width);
 			}
+
+			if (w->fg_shader) {
+				ps->backend_data->ops->set_image_property(
+				    ps->backend_data, IMAGE_PROPERTY_CUSTOM_SHADER,
+				    w->win_image, (void *)w->fg_shader->backend_shader);
+			}
```
Nice catch

yshui commented:
@tryone144 I think I resolved most of the complains :+1:

yshui committed:
backend: add and implement get_shader_attributes

Used for passing back information about whether a shader needs to be
re-rendered every frame.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
core: force redraw if animated shader is in-use

I also noticed there is an opportunity to unify fades, animated shaders,
and the benchmark mode. We should be able to get rid of the fade timer.
A TODO for the future.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
man: define the custom shader interface

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui commented:
@tryone144 i think this PR is complete.

tryone144 reviewed (approved):
LGTM :+1:
Just a question regarding the documentation: Do we want to mention the special value "default" for a shader in the `--help` output as well?

While we are at it, adding both options to the example config-file might be nice.
I've also though about providing a minimal `default.frag` shader as an example:
```glsl
#version 330
in vec2 texcoord;             // texture coordinate of the fragment

//uniform float opacity;        // opacity of the window (0.0 - 1.0)
//uniform float dim;            // dimming factor of the window (0.0 - 1.0, higher means more dim)
//uniform float corner_radius;  // corner radius of the window (pixels)
//uniform float border_width;   // estimated border width of the window (pixels)
//uniform bool invert_color;    // whether to invert the color of the window
uniform sampler2D tex;        // texture of the window
//uniform sampler2D brightness; // estimated brightness of the window, 1x1 texture
//uniform float max_brightness; // configured maximum brightness of the window (0.0 - 1.0)
//uniform float time;           // time in milliseconds, counting from an unspecified starting point

// Default window post-processing:
// 1) invert color
// 2) opacity / transparency
// 3) max-brightness clamping
// 4) rounded corners
vec4 default_post_processing(vec4 c);

// Default window shader:
// 1) fetch the specified pixel
// 2) apply default post-processing
vec4 window_shader() {
    vec4 c = texelFetch(tex, ivec2(texcoord), 0);
    return default_post_processing(c);
}
```

---
The `get_shader_attributes()` interface might be a bit overkill for now, but is definitely the cleaner approach. In contrast to a simple `bool *is_animated` parameter, we use an additional function call into the backend and explicitly store the single attribute in a 64bit state. However, I am not sure if this outweighs the cleaner design...

Theoretically, a shader can make an opaque window (partially) transparent. But it can't think of a sane way to check whether this is actually the case to return this as a stable attribute. This could help in the decision whether to blur the window's background selectively (and not relying on `force-win-blend`). Not sure if the window mode should be set to `WMODE_TRANS` in that case as well. It seems to serve the same purpose.

I can't think of other use-cases that would require backend-specific information about the shader outside of the backend — at least for now (use-cases being: custom textures, multi-file shaders, multiple passes). Maybe for custom vertex shaders, that we can't rely on the window geometry for damage?

tryone144 commented:
Another thing that just came to mind:
While it might be intuitive that redrawing the animated shaders continuously will inevitably increase the CPU load, we could add a short note to the docs.

Are we re-drawing with the screen refresh-rate when vsync is enabled (even on nvidia that has problems with `glFinish()`)? Or do we re-draw as fast as the idle-loop permits?
A short test supports the assumption we draw at the vsync limit (i.e. `vsync = false` results in 100% utilization).

yshui commented on src/backend/backend.h:
```diff
@@ -199,8 +208,24 @@ struct backend_operations {
+	/// Create a shader object from a shader source.
+	///
+	/// Optional
+	void *(*create_shader)(backend_t *backend_data, const char *source)attr_nonnull(1, 2);
```
In fact it would removes the space if I add one :man_shrugging:

yshui commented:
> A short test supports the assumption we draw at the vsync limit (i.e. vsync = false results in 100% utilization).

Hmm, it's an interesting question. But I think this is more or less expected behavior. I tried `glxgears`, and it does the same sort of thing if I disable vsync.

yshui committed:
options: mention the "default" shader

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
picom.sample.conf: add window-shader-fg and window-shader-fg-rule

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
man: include a default shader

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
man: describe where to put a shader file

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

LoganDark commented:
Where can I find documentation on how to use this myself? Can it be used to create rounded corners also, or work at the same time as them? I want to try to use it to implement #824

yshui commented:
@LoganDark see `man picom`. Yes, it's possible to implement rounded corners in your custom shaders.

LoganDark commented:
> 

Will it be possible to separate the post processing passes in the future? Like apply the color adjustments and then my own filter and then the vanilla rounded corners?

yshui commented:
@LoganDark you can copy the [default_post_processing code](https://github.com/yshui/picom/blob/5d6326ae2892718a0d3680f9503364d00dc5859c/src/backend/gl/gl_common.c#L1589) if you want.

LoganDark commented:
I see, it looks like rewriting / copying all the code is the only option, well I suppose this is fine for an initial release, thank you :)

Hope that more modular/granular options are added in the future

LoganDark commented:
Is it possible for a shader to change the opacity of the background-blur? So outside of the window is not blurred?

yshui commented:
@LoganDark No, this option is for custom foreground shaders, hence the `fg` in the name. You can open a feature request for background shaders if you want.

Falkgaard commented:
Hmm, I try to start picom like this: `picom --config ~/.config/picom/picom.conf --window-shader-fg ~/.config/picom/shader.glsl` (the config is the sample one from the source tree, with `backend = "xrender";`), but I get the following:

`[ 08/27/2022 11:25:49.881 get_cfg WARN ] The new window shader interface does not work with the legacy glx backend.`

I've tried explicitly adding a `--backend xrender` arg, but the result is the same.

The picom version is `vgit-e0758` and the shader is:

```glsl
#version 330
in      vec2      texcoord;       // texture coordinate of the fragment
uniform float     opacity;        // opacity of the window (0.0 - 1.0)
uniform float     dim;            // dimming factor of the window (0.0 - 1.0, higher means more dim)
uniform float     corner_radius;  // corner radius of the window (pixels)
uniform float     border_width;   // estimated border width of the window (pixels)
uniform bool      invert_color;   // whether to invert the color of the window
uniform sampler2D tex;            // texture of the window
uniform sampler2D brightness;     // estimated brightness of the window, 1x1 texture
uniform float     max_brightness; // configured maximum brightness of the window (0.0 - 1.0)
uniform float     time;           // time in milliseconds, counting from an unspecified starting point

vec4 default_post_processing( vec4 c );

vec4 window_shader() {
	vec4 c = texelFetch(tex, ivec2(texcoord), 0);
	mix( c, vec4(.5, 1.0, .5, 1.0), .5 ); // just for feedback that the shader works
	return default_post_processing(c);
}
```

Any idea what's going wrong?

tryone144 commented:
> I've tried explicitly adding a `--backend xrender` arg, but the result is the same.

The *OpenGL

Falkgaard commented:
Thank you, with glx it works great! (For some reason I thought the glx backend was the legacy one...)

By the way, do you know if it's possible to make the shader apply to the whole screen (and not specific windows) with the shader conditions? I'm trying to make a CRT-ish shader (scanlines, curvature, etc). I'm using bspwm, if that makes any difference. I tried `--window-shader-fg-rule ~/.config/picom/retro.glsl:'wmwin'`, but that condition seems to be the wrong one for what I'm trying to achieve.

yshui commented:
@Falkgaard I guess it's a bit confusing. There are the legacy backends and the new backends. Each set of backends has its own glx and xrender backend. You were using the legacy backend, because you probably are not using the latest commit and didn't pass `--experiment-backends`.

The new backends are now the default thus if you use the latest git commit you don't need that option anymore.

yshui commented:
> shader apply to the whole screen

This is better achieved with a overall shader, the per-window shader isn't suited for this job.

Falkgaard commented:
I see. Updated to the latest commit. Cheers!

> This is better achieved with a overall shader, the per-window shader isn't suited for this job.

Mm, makes sense. Does Picom support overall shaders or just per-window shaders?

(And if not, if anyone knows of some software that allows this when running a tiling WM, I'd love to hear about it.)

LoganDark commented:
> curvature

Keep in mind this would cause the cursor to no longer be accurate, because picom doesn't render the cursor itself

xplshn commented:
> CRT

Would you mind sharing your shader with us? I've been looking for an accurate CRT shader for PICOM, I found one, but its not accurate at all. I tried porting zfast-crt but failed
